Topics/Tech_Stacks/Spring-Security.md

```diff
@@ -0,0 +1,104 @@
+	- Define application's security constraints in configuration
+		- All Java config: @Configuration
+	- Provides separation of concerns between application code and security
+- Programmatic Security
+	- Spring Security provides an API for custom application coding
+	- Provides greater customization for specific app requirements
+
+### Enabling Spring Security
+1. Edit pom.xml and add `spring-boot-starter-security`
+2. This will automagically secure all endpoints for application
+	- Accessing secured endpoints will prompt for login
+	- Override default user name and generated password in application.properties
+
+- Cross-Site Request Forgery (CSRF)
+	- Spring Security can protect against CSRF attackes
+	- Embed additional authentication data / token into all HTML forms
+	- On subsequent requests, web app will verify token before processing
+	- Primary use case is traditional web applications (HTML forms etc...)
+
+- When to use CSRF Protection
+	- For any normal browser web requests
+	- Traditional web apps with HTML forms to add / modify data
+	- If building REST API for non-browser clients
+		- disable CSRF protection, not required for stateless REST APIs that use POST, PUT, DELETE and / or PATCH
+
+### Database Access
+- Default Spring Security Database Schema
+	- users
+		- username
+		- password
+		- enabled
+	- authorities (same as roles)
+		- username
+		- authority
+
+- BCrypt Encryption
+	- Performs one-way encrypted hashing
+	- Adds a random salt to the password for additional protection
+	- Includes support to defeat brute force attack
+
+	- Generate password
+		- www.luv2code.com/generate-bycrypt-password
+		- Enter plaintext password
+		- Generate a bcrypt password
+		- Multiple runs will generate a different password due to random password salting
+
+	- Login Process
+		1. Retrieve password from db for the user
+		2. Read the encoding algorithm id (bcrypt)
+		3. For case of bcrypt, encrypt plaintext password from login form (using salt from db password)
+		4. Compare encrypted password from login form with encrypted password from db
+		5. If there is a match, login successful
+		6. If no match, login not successful
+	- The password from db is NEVER decrypted because bcrypt is a one-way encryption algorithm
+
+- Security Schema Customization
+	- Tell Spring how to query the custom tables
+	- Provide query to find user by user name
+	- Provide query to find authorities / roles by user name
+
+# MVC Security
+- HTTP Basic Authentication
+- Default login form
+	- Spring Security provides a default login form
+- Custom login form
+	- HTML + CSS
+
+- Bootstrap
+	- Web framework that includes CSS styles and JavaScript
+	- Focused on front-end UI
+
+- Logout
+	- Logout Process
+		- When a logout is processed, by default Spring Security will
+		- Invalidate user's HTTP session and remove session cookies
+		- Send user back to the login page
+		- Append a logout parameter: ?logout
+
+- Restrict URLs Based on Roles
+	- Dev Process
```

Comment:
The majority of the content seems to be conceptual. Please consider adding some practical guide like this Dev Process to the above concepts.

Reply:
Thank you for your insightful feedback! I absolutely agree that blending practical guides with theoretical content can significantly enhance understanding and application. I'm going to work on including step-by-step processes, much like the 'Dev Process' you mentioned, which should make the material more hands-on and user-friendly.